From 4addbafb62db9d7093bb890b43104001c42444b2 Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Tue, 12 Nov 2019 03:09:25 +0200
Subject: [PATCH] [SW-6379] goya2: add interrupt vector

This is needed for all kind of exception
Exceptions includes when CPU is waken up
We assume CPU reset address point to interrupt vector begin
First entry in this vector is the address of __start where execution
shall begin

Change-Id: Iacbdc8404dcc2bb09ff30ce604a272ab6752e1cf
---

diff --git a/arcboot.lds b/arcboot.lds
index a72cd28..17abac3 100644
--- a/arcboot.lds
+++ b/arcboot.lds
@@ -12,26 +12,31 @@
 {
 	ICCM	: ORIGIN = 0x00000000, LENGTH = 64K
 	SRAM	: ORIGIN = 0x10000000, LENGTH = 128M
-	STM		: ORIGIN = 0x18000000, LENGTH = 32M
-	SPI		: ORIGIN = 0x1A000000, LENGTH = 16M
+	STM	: ORIGIN = 0x18000000, LENGTH = 32M
+	SPI	: ORIGIN = 0x1A000000, LENGTH = 16M
 	SBROM	: ORIGIN = 0x1B040000, LENGTH = 64K
 	FWSRAM	: ORIGIN = 0x1B050000, LENGTH = 64K
+	CS	: ORIGIN = 0x1C000000, LENGTH = 64M
 	DCCM	: ORIGIN = 0x80000000, LENGTH = 16K
-	CS		: ORIGIN = 0xf0000000, LENGTH = 64M
 }
 
-REGION_ALIAS("REGION_TEXT",		ICCM)
-REGION_ALIAS("REGION_DATA",		FWSRAM)
+REGION_ALIAS("REGION_IVT",	ICCM)
+REGION_ALIAS("REGION_TEXT",	ICCM)
+REGION_ALIAS("REGION_DATA",	FWSRAM)
 REGION_ALIAS("REGION_SDATA",	DCCM)
-REGION_ALIAS("REGION_SBSS",		DCCM)
-REGION_ALIAS("REGION_BSS",		DCCM)
-REGION_ALIAS("REGION_HEAP",		DCCM)
+REGION_ALIAS("REGION_SBSS",	DCCM)
+REGION_ALIAS("REGION_BSS",	DCCM)
+REGION_ALIAS("REGION_HEAP",	DCCM)
 REGION_ALIAS("REGION_STACK",	DCCM)
 
 ENTRY(__start)
 
 SECTIONS
 {
+	.ivt : {
+		KEEP (*(.ivt));
+	} > REGION_IVT
+
 	.text : {
 		crt0.o (.text*)
 		*(.text)
diff --git a/crt0.S b/crt0.S
index 5e572a3..87a3b76 100644
--- a/crt0.S
+++ b/crt0.S
@@ -13,7 +13,45 @@
 .file "crt0.S"
 
 .extern main
-.text
+
+.section .ivt, "a", @progbits
+
+; Helper macro to define weak symbols to include into interrupt vector table.
+; User code may define those functions in them, so user function will be
+; referenced in the IVT. By default all handlers point to _exit_halt - so they
+; always cause application halt, because if application causes an exception or
+; interrupt, but doesn't set a handler for it - something is wrong in
+; application. Exception is "start" entry of IVT, which points to __start
+; function.
+#define IVT_ENTRY(name) \
+    .word name `\
+    .weak name `\
+    .set name, _exit_halt
+
+; handler's name,		number,	name,			offset in IVT (hex/dec)
+.word __start			; 0	program entry point	0x0		0
+IVT_ENTRY(memory_error)		; 1	memory_error		0x4		4
+IVT_ENTRY(instruction_error)	; 2	instruction_error	0x8		8
+IVT_ENTRY(EV_MachineCheck)	; 3	EV_MachineCheck		0xC		12
+IVT_ENTRY(EV_TLBMissI)		; 4	EV_TLBMissI		0x10		16
+IVT_ENTRY(EV_TLBMissD)		; 5	EV_TLBMissD		0x14		20
+IVT_ENTRY(EV_ProtV)		; 6	EV_ProtV		0x18		24
+IVT_ENTRY(EV_PrivilegeV)	; 7	EV_PrivilegeV		0x1C		28
+IVT_ENTRY(EV_SWI)		; 8	EV_SWI			0x20		32
+IVT_ENTRY(EV_Trap)		; 9	EV_Trap			0x24		36
+IVT_ENTRY(EV_Extension)		; 10	EV_Extension		0x28		40
+IVT_ENTRY(EV_DivZero)		; 11	EV_DivZero		0x2C		44
+IVT_ENTRY(EV_DCError)		; 12	EV_DCError		0x30		48
+IVT_ENTRY(EV_Maligned)		; 13	EV_Maligned		0x34		52
+IVT_ENTRY(EV_Ex14)		; 14	unused			0x38		56
+IVT_ENTRY(EV_Ex15)		; 15	unused			0x3C		60
+IVT_ENTRY(IRQ_Timer0)		; 16	Timer 0			0x40		64
+IVT_ENTRY(IRQ_Timer1)		; 17	Timer 1			0x44		68
+IVT_ENTRY(IRQ_18)		; 18				0x48		72
+IVT_ENTRY(IRQ_19)		; 19				0x4C		76
+IVT_ENTRY(IRQ_20)		; 20				0x50		80
+
+.section .text.__startup, "ax", @progbits
 
 ;; lf2flt: to avoid compiler (buildroot wrapper) warning during linkage
 .global	lf2flt
@@ -21,11 +59,19 @@
 
 .global __start
 .type __start, @function
+.align 4
+
+; Startup code for the ARC600, ARC700 and ARCv2 processors
+; NOTE:  The following restrictions apply on zero overhead loops (other
+; restrictions are not pertinent to this code)
+; - loop end should be 4 instruction words away from the lp_count setting
+;   instruction
+; - loop body should have at least two instruction words
 __start:
 	mov gp, @__SDATA_BEGIN__
 
 	mov_s r2, @__sbss_start	; r2 = start of the bss section
-	sub r3, @_end, r2		; r3 = size of the bss section in bytes
+	sub r3, @_end, r2	; r3 = size of the bss section in bytes
 	; set up the loop counter register to the size (in words) of the bss section
 	asr_s r13, r3
 	asr.f lp_count, r13
@@ -34,21 +80,21 @@
 	mov_s r3, 0
 	; loop to zero out the bss.  Enter loop only if lp_count != 0
 	lpnz @.Lend_zbss
-	st.ab r3, [r2, 4]		; zero out the word
+	st.ab r3, [r2, 4]	; zero out the word
 	nop
 .Lend_zbss:
 
 	;; branch to main
 	bl.d @main
-	mov fp, 0				; initialize frame pointer
-							; r0 contains exit code
+	mov fp, 0		; initialize frame pointer
+				; r0 contains exit code
 	j @_exit_halt
 
 .global _exit_halt
 .type _exit_halt, @function
 .align 4
 _exit_halt:
-	; r0 contains exit code
+				; r0 contains exit code
 	flag 0x01
 	nop
 	nop
